src/lib/y2firstboot/clients/hostname.rb

```diff
@@ -0,0 +1,197 @@
+# frozen_string_literal: true
+
+# Copyright (c) [2020] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+
+Yast.import "UI"
+Yast.import "Arch"
+Yast.import "DNS"
+Yast.import "Host"
+Yast.import "NetworkConfig"
+Yast.import "String"
+Yast.import "Wizard"
+Yast.import "ProductControl"
+Yast.import "ProductFeatures"
+Yast.import "GetInstArgs"
+
+module Y2Firstboot
+  module Clients
+    # Client to set the hostname during first boot
+    #
+    # This is just a renamed version of InstHostnameClient, which was removed from
+    # YaST2-Network when the second stage was removed from the installation
+    # process
+    class Hostname < Yast::Client
+      class << self
+        # @!method valid_dns_proposal=(value)
+        #   @param [Boolean] Whether a valid DNS proposal was done
+        attr_writer :valid_dns_proposal
+
+        # Determines whether a valid DNS proposal was done
+        #
+        # @return [Boolean] Returns true if a DNS proposal was done
+        def valid_dns_proposal
+          @valid_dns_proposal ||= false
+        end
+
+        def run
+          new.run
+        end
+      end
+
+      def initialize
+        textdomain "firstboot"
+      end
+
+      def run
+        Yast.include self, "network/services/dns.rb"
+
+        # only once, do not re-propose if user gets back to this dialog from
+        # the previous screen - bnc#438124
+        if !self.class.valid_dns_proposal
+          Lan.Read(:cache) # handles NetworkConfig too
+          propose_hostname if propose_hostname?
```

Comment:
Probably I would make it single method like `propose_hostname` that handle itself if it need proposal or not.